KIP-73 - Replication Quotas

benstopford opened the pull request:
This applies to Replication Quotas 
based on KIP-73 [(link)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas) originally motivated by KAFKA-1464.

System Tests Run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/544/

**This first PR demonstrates the approach**. 

**_Overview of Change_**
The guts of this change are relatively small. Throttling occurs on both leader and follower sides. A single class tracks the throttled throughput in and out of each broker (**_ReplicationQuotaManager_**).  

On the follower side, the Follower Throttled Rate is calculated as fetch responses arrive. Then, before the next fetch request is sent, we check to see if the quota is violated, removing throttled partitions from the request if it is. This is all encapsulated in a few lines of code in the **_ReplicaFetcherThread_**. There is existing code to handle temporal back off, if the request ends up being empty. 

On the leader side it's a little more complex. When a fetch request arrives in the leader, it is built, partition by partition, in **_ReplicaManager.readFromLocalLog_**. As we put each partition into the fetch response, we check if the total size fits in the current quota. If the quota is exceeded, the partition will not be added to the fetch response. Importantly, we don't increase the quota at this point, we just check to see if the bytes will fit. 

Now, if there aren't enough bytes to send the response immediately, which is common if we're catching up and throttled, then the request will be put in purgatory. I've added some simple code to **_DelayedFetch_** to handle throttled partitions (throttled partitions are checked against the quota, rather than the messages available in the log).

When the delayed fetch completes, and exits purgatory, _**ReplicaManager.readFromLocalLog**_ will be called again. This is why _**ReplicaManager.readFromLocalLog**_ does not actually increase the quota, it just checks whether enough bytes are available for a partition. 

Finally, when there are enough bytes to be sent, or the delayed fetch times out, the response will be sent. Before it is sent the throttled-outbound-rate is increased, based on the size of throttled partitions being sent. This is at the end of _**KafkaApis.handleFetchRequest**_, exactly where client quotas are recorded. 

There is an acceptance test which asserts the whole throttling process stabilises on the desired value. This covers a number of use cases including many-to-many replication. See **_ReplicationQuotaTest_**.

Note: 
It should be noted that this protocol can over-request. The request is built, based on the quota at time t1 (_ReplicaManager.readFromLocalLog_). The bytes in the response are recorded at time t2 (end of _KafkaApis.handleFetchRequest_), where t2 > t1. For this reason I originally included an OverRequestedRate as a JMX metric, but testing has not seen revealed any obvious issue. Over-requesting is quickly compensated by subsequent requests, stabilising close to the quota value. 

_**Main stuff left to do:**_
- The fetch size is currently unbounded. This will be addressed in KIP-74, but we need to ensure this ensures requests don’t go beyond the throttle window.
- There are two failures showing up in the system tests on this branch:  StreamsSmokeTest.test_streams (which looks like it fails regularly) and OffsetValidationTest.test_broker_rolling_bounce (which I need to look into)

_**Stuff left to do that could be deferred:**_ 
- Add the extra metrics specified in the KIP. 
- There are no system tests. 
- There is no validation for the cluster size / throttle combination that could lead to ISR dropouts

ijuma commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,57 @@
+
+object QuotaFactory {
+
+  object QuotaType extends Enumeration {
```
The Scala `Enumeration` class has a bunch of problems and we typically use ADTs. An example is `RackAwareMode`.

benstopford replied:
Changed - thanks bud.

benstopford committed:
KAFKA-1464: All replication quotas changes up to "Add dynamic broker configuration and apply it to the property replication.quota.throttled.rate"

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:
```diff
@@ -112,19 +112,42 @@ public void record(double value, long timeMs) {
+
+    /**
+     * Evaluate whether the value passed causes the quota to be exceeded on any metric
+     * @param delta A delta to be added to each metric before it is evalutated
```
delta is no longer valid. Typo evalutated

ijuma replied:
I think Ben just meant the `value` parameter to be called `delta`. From an API perspective, it seems like it would make sense to also have a public `checkQuota` method if we expose a `checkQuotaWithDelta`.

benstopford replied:
Renamed value->delta. Added an (unused) method: public void checkQuotas()

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/FixedWindowRate.java:
```diff
@@ -0,0 +1,45 @@
+ * early in a window, they will be represented over the full
+ * window's duration.
+ */
+public class FixedWindowRate extends Rate {
```
Is that only used for testing? If so, could we define it under src/test?

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/stats/SimpleRate.java:
```diff
@@ -0,0 +1,32 @@
+public class SimpleRate extends Rate {
+
+    @Override
+    public long windowSize(MetricConfig config, long now) {
```
Do we need this new class? Rate.windowSize() is modified this way to address KAFKA-2443 and KAFKA-2567. It would be better if all sensors are of the same type of Rate.

benstopford replied:
I've consolidated onto a single class, keeping SimpleRate. This encompasses a different approach to fixing KAFKA-2567 whilst being a little simpler to test.

junrao replied:
Having two different rates is going to make it harder for developers to decide which one to use. If this is strictly better than Rate, perhaps we should just change Rate.windowSize(). If this is just for testing, perhaps we can create SimpleRate in test?

junrao commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -185,13 +187,15 @@ object LogConfig {
-    "if the difference in timestamp exceeds this threshold. This configuration is ignored if message.timestamp.type=LogAppendTime."  
+    "if the difference in timestamp exceeds this threshold. This configuration is ignored if message.timestamp.type=LogAppendTime."
+  val ThrottledReplicasListDoc = "A list of replicas for which log replication should be throttled. The list should describe a set of " +
+    "replicas in the form [PartitionId]-[BrokerId]:[PartitionId]-[BrokerId]:..."
```
KIP says using comma separated.

benstopford replied:
Our standard mechanism won't permit commas in config, hence I proposed this. I did actually change the KIP.

ijuma replied:
Not sure I understand, we support multiple items separated by commas in configs via the `LIST` type.

benstopford replied:
Dynamic configs don't support commas. I've added a task to look at changing this.

ijuma replied:
@benstopford they should do, see `cleanup.policy` for an example. There's a test in `LogCleanerIntegrationTest.testCleansCombinedCompactAndDeleteTopic` that verifies this.

benstopford replied:
@ijuma That test verifies you can create a LogConfig object, but doesn't verify that you can update that same log config from the ConfigCommand, which you can't currently.

ijuma replied:
Interesting, this is an issue for KIP-71 too then. cc @dguy

dguy replied:
Hmm, yes indeed. It appears ConfigCommand doesn't like commas. TopicCommand OTOH does.

benstopford replied:
@ijuma @dguy   I've added support for comma separated lists in the ConfigCommand. You specify the list using a square bracket: k1=v1,k2=[v2,v3]

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -63,6 +64,29 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
+      allReplicas
+    else
+      configValue
+        .split(":")
```
The KIP says it's comma separated.

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -63,6 +64,29 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
+    else
+      configValue
+        .split(":")
+        .filter(_.split("-")(1).toInt == brokerId) //match replica
```
Perhaps we can avoid calling split("-") twice by doing that once first?

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+  */
+class BrokerConfigHandler(private val config: KafkaConfig, private val quotaManagers: QuotaManagers) extends ConfigHandler with Logging{
+
+  def processConfigChanges(proposedConfigChangeBrokerId: String, proposedProperties: Properties) {
```
To be consistent with the existing naming, perhaps the params can be brokerId and brokerConfig?

benstopford replied:
OK

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+class BrokerConfigHandler(private val config: KafkaConfig, private val quotaManagers: QuotaManagers) extends ConfigHandler with Logging{
+
+  def processConfigChanges(proposedConfigChangeBrokerId: String, proposedProperties: Properties) {
+    if(config.brokerId == proposedConfigChangeBrokerId.trim.toInt) {
```
space after if

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+    if(config.brokerId == proposedConfigChangeBrokerId.trim.toInt) {
+      val limit = proposedProperties.getProperty(KafkaConfig.ThrottledReplicationRateLimitProp).toLong
+      config.mutateConfig(KafkaConfig.ThrottledReplicationRateLimitProp, limit)
+      if (proposedProperties.containsKey(KafkaConfig.ThrottledReplicationRateLimitProp)) {
```
Shouldn't we do this check before line 127?

benstopford replied:
Yes, good spot. Thank you.

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+      val limit = proposedProperties.getProperty(KafkaConfig.ThrottledReplicationRateLimitProp).toLong
+      config.mutateConfig(KafkaConfig.ThrottledReplicationRateLimitProp, limit)
+      if (proposedProperties.containsKey(KafkaConfig.ThrottledReplicationRateLimitProp)) {
+        quotaManagers.leaderReplication.updateQuota(new Quota(limit, true))
```
It's clearer if we use Quota.upperBound(limit) in this and the next line.

benstopford replied:
Agreed. Changed.

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+    }
+  }
+
+  def isValid(value: String): Boolean = {
```
This can be private?

benstopford replied:
Yeah - it was just used by a test. Have changed to test the ensureValid method instead.

junrao commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -111,7 +116,7 @@ class DelayedFetch(delayMs: Long,
 
     // Case D
-    if (accumulatedSize >= fetchMetadata.fetchMinBytes)
+    if (accumulatedSize >= fetchMetadata.fetchMinBytes || (containsThrottledPartitions && !quota.isQuotaExceededBy(0)))
```
Does this guarantee that the minBytes requirement is met? For example, suppose that all replicas are throttled and quota is not exceeded at this point. However, when we call forceComplete(0, we could get fewer bytes than minBytes and return the fetch response prematurely.

Same question in case C on line 94. If the fetch offset is on an old segment from the last segment, we force a return immediately. However, it could be that the quota is violated and we will return an empty response.

benstopford replied:
Will change to quota.isQuotaExceededBy(fetchMetadata.fetchMinBytes).

junrao commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -37,6 +37,7 @@ import org.I0Itec.zkclient.{IZkStateListener, IZkChildListener, ZkClient}
 object ConfigType {
   val Topic = "topics"
   val Client = "clients"
+  val Broker: String = "broker"
```
To be consistent, should broker be brokers?

ijuma replied:
Nitpick: we don't need the `String` type annotation.

benstopford replied:
Done. thanks

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -144,7 +145,7 @@ class KafkaApis(val requestChannel: RequestChannel,
       val leaderAndIsrResponse =
         if (authorize(request.session, ClusterAction, Resource.ClusterResource)) {
-          val result = replicaManager.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest, metadataCache, onLeadershipChange)
+          val  result = replicaManager.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest, metadataCache, onLeadershipChange)
```
No need to change this line?

benstopford replied:
ok

junrao replied:
extra space before result

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -506,10 +506,32 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  def sizeOfThrottledPartitions(fetchRequest: FetchRequest, mergedPartitionData: Map[TopicAndPartition, FetchResponsePartitionData], quota: ReplicationQuotaManager): Int = {
```
This can be private?

benstopford replied:
sure can

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -517,8 +524,12 @@ object KafkaConfig {
+  val NumReplicationQuotaSamplesDoc = "The number of samples to retain in memory for replication quotas"
+  val QuotaWindowSizeSecondsDoc = "The time span of each sample for client quotas"
+  val ReplicationQuotaWindowSizeSecondsDoc = "The time span of each sample for replication quotas"
+  val ThrottledReplicationRateLimitDoc = "An long representing an upper bound on replication traffic for replicas enumerated in the " +
```
An => A

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -517,8 +524,12 @@ object KafkaConfig {
+  val QuotaWindowSizeSecondsDoc = "The time span of each sample for client quotas"
+  val ReplicationQuotaWindowSizeSecondsDoc = "The time span of each sample for replication quotas"
+  val ThrottledReplicationRateLimitDoc = "An long representing an upper bound on replication traffic for replicas enumerated in the " +
+    "property quota.replication.throttled.replicas. This property can be set dynamically."
```
Could we use the defined name for quota.replication.throttled.replicas?

benstopford replied:
good idea. thanks.

junrao commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,62 @@
+package kafka.server
```
Missing license header.

benstopford replied:
added. thanks

junrao commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,62 @@
+  case object Fetch extends QuotaType
+  case object Produce extends QuotaType
+  case object LeaderReplication extends QuotaType
+  case object  FollowerReplication extends QuotaType
```
extra space after object

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,62 @@
+object QuotaFactory {
+
+
+  object UnbreakableQuota extends ReadOnlyQuota {
```
Perhaps NoQuota is better?

ijuma replied:
`UnboundedQuota` maybe?

benstopford replied:
UnboundedQuota sounds good to me.

junrao commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,62 @@
+    override def isQuotaExceededBy(bytes: Int): Boolean = false
+  }
+
+  case class QuotaManagers(client: Map[QuotaType, ClientQuotaManager], leaderReplication: ReplicationQuotaManager, followerReplication: ReplicationQuotaManager)
```
For clarity, would it be better to rename leaderReplication and followerReplication to leaderQuotaManager and followerQuotaManager?

ijuma replied:
Given that the containing class is `QuotaManagers`, it seems like it may be OK to not repeat `QuotaManager` for each field. If we do rename it as per Jun's suggestion, then we probably should rename `client` too.

ijuma replied:
This class seems a bit inconsistent in that it groups the client replication types in a map, but not the replication ones. It seems like it might be better to either have them all as fields or all in maps. Is there a reason to do it this way instead?

benstopford replied:
Yeah - I should have put a comment on that. It's a historical artefact from the way the original client quota managers worked. I knew I had to rework it. 

It should be a bit better now.

junrao commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,62 @@
+
+    val clientQuotaManagers = Map[QuotaType, ClientQuotaManager](
+      QuotaType.Produce ->
+        new ClientQuotaManager(producerQuotaManagerCfg, metrics, QuotaType.Produce.toString, new org.apache.kafka.common.utils.SystemTime),
```
No need for the package name org.apache.kafka.common.utils. Ditto two lines below.

benstopford replied:
done

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -267,17 +267,30 @@ class ReplicaFetcherThread(name: String,
     new FetchRequest(new JFetchRequest(replicaId, maxWait, minBytes, requestMap.asJava))
   }
 
+  override def fetchResponseProcessingComplete(responseData: Map[TopicAndPartition, PD]) = {
```
Do we really need fetchResponseProcessingComplete()? Could we just do the logic here in processPartitionData?

benstopford replied:
Nope. Hangover from when we had separate throttled replica fetcher threads. Removed.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -489,33 +494,48 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  def check(totalThrottledBytes: Int, bound: Int): Int = {
```
Is this still needed? If not, it seems that we don't need to expose bound() as a public api in ReplicationQuotaManager.

benstopford replied:
Nope. Removed.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -489,33 +494,48 @@ class ReplicaManager(val config: KafkaConfig,
       BrokerTopicStats.getBrokerTopicStats(topic).totalFetchRequestRate.mark()
       BrokerTopicStats.getBrokerAllTopicsStats().totalFetchRequestRate.mark()
 
+      logger.info("starting readFromLocalLog for "+ readPartitionInfo.map(_._1))
```
Does this need to be info? Also, don't need to reference logger. There are a few other places like that.

benstopford replied:
I'll fix all logging etc when I have something I'm prepared to merge. This is just a first cut remember.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -536,10 +556,22 @@ class ReplicaManager(val config: KafkaConfig,
+              //Try the read first, this tells us whether we need all of adjustedFetchSize for this partition
+              var fetch = log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //If the read for this partition caused the quota to be excedded then zero it out, so it is excluded
```
typo excedded

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -536,10 +556,22 @@ class ReplicaManager(val config: KafkaConfig,
+
+              //If the read for this partition caused the quota to be excedded then zero it out, so it is excluded
+              if(quota.isThrottled(TopicAndPartition(topic, partition))) {
+                if (quota.isQuotaExceededBy(check(totalThrottledBytesInFetch, quota.bound())))
```
Hmm, shouldn't we pass in fetch.messageSet.sizeInBytes to quota.isQuotaExceededBy()?

benstopford replied:
Sorry - deleted previous. As I understand it the fetch, here, is just a single partition's data. Thus I'm keeping a running total of throttled bytes in all partitions, checking against the quota to see when the 'proposed' fetch will exceed to. Remember isQuotaExceededBy() doesn't record anything. Let me know if I'm missing something.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -536,10 +556,22 @@ class ReplicaManager(val config: KafkaConfig,
+              //If the read for this partition caused the quota to be excedded then zero it out, so it is excluded
+              if(quota.isThrottled(TopicAndPartition(topic, partition))) {
+                if (quota.isQuotaExceededBy(check(totalThrottledBytesInFetch, quota.bound())))
+                  fetch = log.read(offset, 0, maxOffsetOpt)
```
Do we have to read from the log again? It seems that we can just set fetch.messageSet to an empty ByteBufferMessageSet.

benstopford replied:
yes - that's better.

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+import kafka.server.Constants._
+import kafka.server.ReplicationQuotaManagerConfig._
+import kafka.utils.CoreUtils._
+import kafka.utils.{CoreUtils, Logging}
```
unused import CoreUtils

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+  val InactiveSensorExpirationTimeSeconds = 3600
+}
+
+trait ReadOnlyQuota {
```
ReadOnlyQuota seems a bit too general. Perhaps sth like ReplicaQuota?

benstopford replied:
OK

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+}
+
+trait ReadOnlyQuota {
+  def bound(): Int
```
bound() => upperBound() to make it clear?

benstopford replied:
good idea.

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+  private val lock = new ReentrantReadWriteLock()
+  private val throttledPartitions = new ConcurrentHashMap[String, Seq[Int]]()
+  private var quota: Quota = null
+  private val sensor = createSensor(quota, replicationType)
```
This may not be enough since the sensor can be removed after it's inactive for some time (say, throttled replicas are removed). When that happens, sensor will be pointing to an obsolete object. To be safe, we probably need a getOrCreateQuotaSensors() method like in ClientQuotaManager.

benstopford replied:
That makes sense. Will change. Thanks for the heads up.

benstopford replied:
Have extracted the logic from the ClientQuotaManager and reused, which, I think, is better!

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+  private var quota: Quota = null
+  private val sensor = createSensor(quota, replicationType)
+
+  private def rateMetricName(): MetricName = metrics.metricName("byte-rate", replicationType, s"Tracking byte-rate for $replicationType")
```
Could this be a val instead of a method?

benstopford replied:
sure can.

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+    *
+    * @return
+    */
+  def getQuota(): Quota = {
```
It seems that this can be private? Also, getQuota() can be just quota().

benstopford replied:
yes. thanks

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+      sensor.checkQuotaWithDelta(value)
+    } catch {
+      case qve: QuotaViolationException =>
+        logger.info("%s: Quota violated for sensor (%s), metric: (%s), metric-value: (%f), bound: (%f), proposedValue (%s)".format(replicationType, sensor.name(), qve.metricName, qve.value, qve.bound, value))
```
Does this need to be at info? Also, not need to use logger. info() is enough.

benstopford replied:
will do all these at the end.

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,219 @@
+    *
+    * @return
+    */
+  override def bound(): Int = {
```
Hmm, should bound() return int? With things like infiniband, it actually can be legit to set a quota larger than 2GB/sec.

benstopford replied:
No should be long everywhere. Will change.

benstopford replied:
this was changed

benstopford commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -536,10 +556,22 @@ class ReplicaManager(val config: KafkaConfig,
+
+              //If the read for this partition caused the quota to be excedded then zero it out, so it is excluded
+              if(quota.isThrottled(TopicAndPartition(topic, partition))) {
+                if (quota.isQuotaExceededBy(check(totalThrottledBytesInFetch, quota.bound())))
```
No - it should span the whole fetch, which it did, but i've made a refactoring error. Will fix.

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/QuotaViolationException.java:
```diff
@@ -17,15 +17,26 @@
     private static final long serialVersionUID = 1L;
+    public MetricName metricName;
+    public Double value;
+    public Double bound;
```
These should be final and probably exposed via an accessor.

benstopford replied:
I wouldn't typically add accessors here, although I would have some years back. If you're super keen on this kind of stuff I'll change.

ijuma replied:
I'm not super-keen on the accessors, it's just the general Kafka convention. I'm keen on the fields being `final` though. Also, do we want `value` and `bound` to be nullable? If not, then they should be lowercase `double`.

benstopford replied:
ok - changed

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -1,45 +1,46 @@
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
```
Not sure if these formatting changes are intended.

benstopford replied:
nope. removed

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -63,6 +64,29 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
       logs.foreach(_.config = logConfig)
     }
+
+    val brokerId: Int = kafkaConfig.brokerId
```
Nitpick: we don't usually include the type annotation for simple local variables like this.

benstopford replied:
good spot. thank you.

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -63,6 +64,29 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
+      val partitions = parseThrottledPartitions(topicConfig, brokerId)
+      quotas.leaderReplication.markReplicasAsThrottled(topic, partitions)
+      quotas.followerReplication.markReplicasAsThrottled(topic, partitions)
+      logger.info(s"Setting throttled partitions on broker $brokerId to ${partitions.map(_.toString)}")
```
It seems to me that the `partitions.map(_.toString)` doesn't do anything since the default behaviour is to call `toString` on each element anyway.

benstopford replied:
Cool. thanks.

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+      throw new ConfigException(name, value, s"$name  must be a string")
+    if (!isValid(value.toString)) {
+      throw new ConfigException(name, value, s"$name  must match for format [number]-[number]:[number]-[number]:[number]-[number] etc")
+    }
```
Generally in Scala, you would write this like:

``` scala
value match {
  case s: String => ...
  case _ => throw new ConfigException(...)
```

benstopford replied:
Changed

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +96,55 @@ object ClientConfigOverride {
+  def processConfigChanges(proposedConfigChangeBrokerId: String, proposedProperties: Properties) {
+    if(config.brokerId == proposedConfigChangeBrokerId.trim.toInt) {
+      val limit = proposedProperties.getProperty(KafkaConfig.ThrottledReplicationRateLimitProp).toLong
+      config.mutateConfig(KafkaConfig.ThrottledReplicationRateLimitProp, limit)
```
Can we not make a copy of KafkaConfig instead?

benstopford replied:
Interesting. So it's not actually used anywhere. The value is just passed into the quota managers, following the same pattern used in the ClientQuotaManager. So maybe we don't need to change KafkaConfig at all? What do you think?

ijuma commented on core/src/main/scala/kafka/server/QuotaFactory.scala:
```diff
@@ -0,0 +1,62 @@
+    )
+
+    val leader = new ReplicationQuotaManager(replicationQuotaManagerCfg, metrics, QuotaType.LeaderReplication.toString, new SystemTime)
+    val follower = new ReplicationQuotaManager(replicationQuotaManagerCfg, metrics, QuotaType.FollowerReplication.toString, new SystemTime)
```
It would be nice if a `Time` instance was passed instead of hardcoding `SystemTime` here.

benstopford replied:
OK

ijuma commented on core/src/test/scala/unit/kafka/server/ReplicaManagerQuotasTest.scala:
```diff
@@ -0,0 +1,152 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package unit.kafka.server
```
There should be no `unit.` here.

benstopford replied:
thanks

ijuma commented on core/src/test/scala/unit/kafka/server/ReplicationQuotaTest.scala:
```diff
@@ -0,0 +1,211 @@
+class ReplicationQuotaTest extends ZooKeeperTestHarness {
+  def TenPercentError(value: Int): Int = Math.round(value * 0.1).toInt
+
+  def FifteenPercentError(value: Int): Int = Math.round(value * 0.15).toInt
```
Nitpick: methods should start with lowercase letter.

benstopford replied:
ok

benstopford committed:
KAFKA-1464: Altered ReassignPartitionsCommand to accept throttle (also changes behaviour to default to blocking to completion)

benstopford committed:
KAFKA-1464: Added first version of asymetric throttle tests

benstopford committed:
KAFKA-1464: Changed asymetric tests to be 6-Leaders => 2-Followers

benstopford committed:
KAFKA-1464: Added todo

benstopford committed:
KAFKA-1464: Tidied Sensor based on Ismael's & Jun's feedback

benstopford committed:
KAFKA-1464: Wired in the changing of Broker properties dynamically using the ConfigCommand (kafka-configs.sh).

benstopford committed:
KAFKA-1464: added todo

benstopford committed:
KAFKA-1464: Small change suggested by Jun

benstopford committed:
KAFKA-1464: More small changes suggested by Jun

benstopford committed:
KAFKA-1464: More small changes suggested by Jun/Ismael

benstopford committed:
KAFKA-1464: Refactored the QuotaFactory so it enumerates all the quota managers rather than using a map based on API key (as was used in the previous implementation). No functional change.

benstopford committed:
KAFKA-1464: Rename only: changed the names of the various quota managers

benstopford committed:
KAFKA-1464: Removed fetchResponseProcessingComplete() method which was only needed when we had separate throttled replica fetcher threads.

ijuma commented on core/src/test/scala/unit/kafka/server/ReplicationQuotaTest.scala:
```diff
@@ -0,0 +1,211 @@
+  def waitForOffset(topicAndPart: TopicAndPartition, offset: Int, servers: Seq[KafkaServer] = brokers): Boolean = {
+    var result = true
+    result = result && servers.forall { item =>
+      offset == (if (item.getLogManager().getLog(topicAndPart) == None) 0 else item.getLogManager().getLog(topicAndPart).get.logEndOffset)
```
You can express the the RHS of `==` as:

``` scala
item.getLogManager.getLog(topicAndPart).map(_.logEndOffset).getOrElse(0)
```

benstopford replied:
Ah yes. Thanks

ijuma commented on core/src/test/scala/unit/kafka/server/ReplicationQuotaTest.scala:
```diff
@@ -0,0 +1,211 @@
+  }
+
+  def waitForOffset(topicAndPart: TopicAndPartition, offset: Int, servers: Seq[KafkaServer] = brokers): Boolean = {
+    var result = true
```
Is this a left-over debugging thing? It seems like you don't need this var at all.

benstopford replied:
It most certainly is. Thank you. I'll do a refactor of these tests in my final pass.

benstopford committed:
KAFKA-1464: Rename QuotasTest -> ClientQuotasTest

benstopford committed:
KAFKA-1464: minor rename

benstopford committed:
KAFKA-1464:
(a) Changed Int=>Long where bound is used
(b) Removed check which ensures we don't push quota negative (added test also). It should be ok to evaluate a quota that leads to a negative bound. It will just fail the acceptable() call
(c) Tidied the way we zero out bytes when the quota is exceeded
(d) Refactored ReplicationQuotaManager to remove calls only used in tests/logging (including bound, rate and getQuota)

benstopford committed:
KAFKA-1464:  altered delayed operation to compare against accumulated size, rather than just fetchMinBytes

benstopford committed:
KAFKA-1464: Change to leader throttle algorithm: Ditched the isExceededBy(bytes) approach in favour of simply excluding partitions if the quota is exceeded. The implication is we will always send whatever throttled bytes are available when we are below quota (assuming there are minFetchBytes).

benstopford committed:
KAFKA-1464: Small sylistic changes based on Jun's feedback

benstopford committed:
KAFKA-1464: Extracted Sensor Access logic from the ClientQuotaManager and reused in the ReplicationQuotaManager

benstopford committed:
KAFKA-1464: Comment only

benstopford committed:
KAFKA-1464: Stylistic changes only

benstopford committed:
KAFKA-1464: Stylistic changes

benstopford committed:
KAFKA-1464: Stylistic changes

benstopford committed:
KAFKA-1464: More stylistic changes

benstopford committed:
KAFKA-1464: More stylistic changes

benstopford committed:
KAFKA-1464: Getters and setters

benstopford committed:
KAFKA-1464: Joined the time interfaces and pushed time from QuotaFactory up to KafkaServer

benstopford committed:
KAFKA-1464: formatting

benstopford committed:
KAFKA-1464: Removed FixedWindowRate. Changed Simple rate to use a fixed first window.

ijuma commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -103,11 +108,60 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
   }
+
+  def inThread()(fun: => Unit) = {
+    Utils.newThread("ReplicationTestHelperThread", CoreUtils.runnable(fun), false).start
```
We are leaking the thread right? If you just want to execute something in the background, you can do something like:

``` scala
import scala.concurrent.{Future, ExecutionContext}
import ExecutionContext.Implicits._
...
Future(codeToRunInTheGlobalThreadPool)
```

If you store the future, you can also await on the result or just completion by using `Await.result` or `Await.ready` (probably not applicable in this case, but worth knowing).

benstopford replied:
thanks. good to know.

ijuma replied:
Thanks for changing the calling code. Can we remove this method then?

benstopford replied:
done. thanks.

benstopford committed:
KAFKA-1464: Ensure that Throttled Configs can be deleted / reset.

benstopford committed:
KAFKA-1464: Stylistic change

benstopford committed:
KAFKA-1464: Reverted mutability of KafkaConfig

benstopford committed:
KAFKA-1464: Consolidated the quotas acceptance tests into a single all-encompassing test.

benstopford committed:
KAFKA-1464: Cleaned up logging

ijuma commented on clients/src/main/java/org/apache/kafka/common/metrics/QuotaViolationException.java:
```diff
@@ -17,15 +17,38 @@
 
-    public QuotaViolationException(String m) {
-        super(m);
+    public QuotaViolationException(MetricName metricName, Double value, Double bound) {
```
Nitpick: `value` and `bound` should be lowercase `double` since they can't be null (annoying that Scala and Java are different in this respect so it's a bit confusing when writing code in both languages).

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -41,7 +41,7 @@ object ConfigCommand {
 
     if(args.length == 0)
-      CommandLineUtils.printUsageAndDie(opts.parser, "Add/Remove entity (topics/clients) configs")
+      CommandLineUtils.printUsageAndDie(opts.parser, "Add/Remove entity (topics/clients/broker) configs")
```
Nitpick: should be `brokers` to be consistent with the other ones? Or should the other ones be made singular?

benstopford replied:
Good point. I've changed to "Add/Remove entity config for a topic, client or broker" (command only lets you change one at a time)

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -97,6 +97,15 @@ object ConfigCommand {
 
+  private def parseBroker(broker: String): Seq[Int] = {
+    try {
+      return Seq(broker.toInt)
```
Nitpick: no return needed and no blocks are needed. Example:

``` scala
try broker.toInt
catch {
  ...
}
```

benstopford replied:
thanks

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -97,6 +97,15 @@ object ConfigCommand {
+      return Seq(broker.toInt)
+    }catch {
+      case e: NumberFormatException =>
+        throw new IllegalArgumentException("The broker's Entity Name must be a single integer value")
```
I'd include the `broker` variable in the message.

benstopford replied:
good idea. thanks

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -97,6 +97,15 @@ object ConfigCommand {
     }
   }
 
+  private def parseBroker(broker: String): Seq[Int] = {
```
I think it would be a bit better if this method returned an `Int` and the caller just wrapped the result in `Seq(...)`.

benstopford replied:
ok

ijuma commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -190,10 +200,9 @@ object ConfigCommand {
-      if(! entityTypeVal.equals(ConfigType.Topic) && ! entityTypeVal.equals(ConfigType.Client)) {
-        throw new IllegalArgumentException("--entity-type must be '%s' or '%s'".format(ConfigType.Topic, ConfigType.Client))
+      if(! entityTypeVal.equals(ConfigType.Topic) && ! entityTypeVal.equals(ConfigType.Client) && ! entityTypeVal.equals(ConfigType.Broker)) {
+        throw new IllegalArgumentException("--entity-type must be '%s' or '%s' or '%s'".format(ConfigType.Topic, ConfigType.Client, ConfigType.Broker))
```
Seems like this code would be a bit nicer if we had a `val supportedTypes = Set(ConfigType.Topic, ConfigType.Client, ConfigType.Broker)` somewhere.

benstopford replied:
definitely!

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -63,6 +66,32 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
+        .filter(_(1).toInt == brokerId //Filter this replica
+        )
+        .map(_(0).toInt).toSeq //convert to list of partition ids
+    }
```
Maybe something like:

``` scala
configValue.trim match {
  case "" => ...
  case "*" => ...
  case _ => ...
}
```

I think you also need to handle errors if the format doesn't match what you expect, right? At the moment, we will get unhelpful `ArrayOutOfBoundsException`s and `NumberFormatException`s.

benstopford replied:
Yep. that's the idea.

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -63,6 +66,32 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
+        .filter(_(1).toInt == brokerId //Filter this replica
+        )
+        .map(_(0).toInt).toSeq //convert to list of partition ids
+    }
```
I see that there's a `ThrottledReplicaValidator`, is that ensuring that things will be in the right format by the time we get here?

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +101,50 @@ object ClientConfigOverride {
+object ThrottledReplicaValidator extends Validator {
+  override def ensureValid(name: String, value: scala.Any): Unit = {
+    value match {
+      case s:String => if (!isValid(value))
```
Nitpick: space after `:`

ijuma commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +101,50 @@ object ClientConfigOverride {
+    }
+  }
+
+  private def isValid(value: scala.Any): Boolean = {
```
Unless I am missing something, I think this should take a `String` since the caller is passing it a `String`. Then you don't need `toString` below.

benstopford replied:
it just means the tostring has to be called in the ensureValid method, so six of one and half a dozen of the other. unless I'm missing something.

ijuma replied:
No, you don't need to do that because of the pattern matching clause. You have to pass `s` instead of `value`.

benstopford replied:
ah, gottcha. thanks

ijuma commented on core/src/main/scala/kafka/server/DynamicConfigManager.scala:
```diff
@@ -92,6 +93,7 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
             case Some(ConfigType.Topic) => ConfigType.Topic
             case Some(ConfigType.Client) => ConfigType.Client
+            case Some(ConfigType.Broker) => ConfigType.Broker
             case _ => throw new IllegalArgumentException("Config change notification must have 'entity_type' set to either 'client' or 'topic'." +
```
This message doesn't mention `broker`. Another case where having a definition of supported types would make the code more robust.

benstopford replied:
done (with enumerated list)

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1047,9 +1031,8 @@ class KafkaApis(val requestChannel: RequestChannel,
-      quotaManager.shutdown()
-    }
+    quotas.produce.shutdown()
+    quotas.fetch.shutdown()
```
Wouldn't it be better to have a `shutdown()` in `quotas` that closes all of them?

benstopford replied:
ok.

ijuma commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -771,6 +786,7 @@ object KafkaConfig {
 }
 
 class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean) extends AbstractConfig(KafkaConfig.configDef, props, doLog) {
+  private val lock = new ReentrantReadWriteLock()
```
It doesn't seem like this is used any more.

benstopford replied:
thanks

ijuma commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -559,6 +570,8 @@ object KafkaConfig {
   val SaslKerberosMinTimeBeforeReloginDoc = SaslConfigs.SASL_KERBEROS_MIN_TIME_BEFORE_RELOGIN_DOC
   val SaslKerberosPrincipalToLocalRulesDoc = SaslConfigs.SASL_KERBEROS_PRINCIPAL_TO_LOCAL_RULES_DOC
 
+  def mutableConfigs = Seq(KafkaConfig.ThrottledReplicationRateLimitProp)
```
Do we still need this?

benstopford replied:
yep. it's the definitive list of broker configs you can change.

ijuma commented on core/src/main/scala/kafka/utils/CoreUtils.scala:
```diff
@@ -277,4 +279,10 @@ object CoreUtils extends Logging {
     listenerList.map(listener => EndPoint.createEndPoint(listener)).map(ep => ep.protocolType -> ep).toMap
   }
+
+  def wrapInProperties(key: String, value: String) = {
```
Maybe this should be `def props(map: Map[String, String])` and then you would use it like:

``` scala
CoreUtils.props(Map("foo" -> "blah"))
```

benstopford replied:
have removed & refactored original. was a bit pointless.

ijuma commented on core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala:
```diff
@@ -0,0 +1,57 @@
+    val producer = TestUtils.createNewProducer(TestUtils.getBrokerListStrFromServers(servers), retries = 5, acks = 0)
+    for (x <- 0 until num)
+      producer.send(new ProducerRecord(topic, new Array[Byte](size))).get
+  }
```
Have you seen `TestUtils.produceMessages`?

benstopford replied:
Have changed

ijuma commented on core/src/test/scala/unit/kafka/admin/TestAdminUtils.scala:
```diff
@@ -0,0 +1,28 @@
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package unit.kafka.admin
```
`unit.` should not be here.

benstopford replied:
thanks

benstopford committed:
KAFKA-1464: Double->double

benstopford committed:
KAFKA-1464: change plurality of comment => "Add/Remove entity config for a topic, client or broker"

benstopford committed:
KAFKA-1464: Collection of small changes based on Ismael's feedback

benstopford committed:
KAFKA-1464: Small stylistic change

benstopford committed:
KAFKA-1464: Remove redundant method

benstopford committed:
KAFKA-1464: Tidied exception raising

benstopford committed:
KAFKA-1464: small commit error

benstopford committed:
KAFKA-1464: Add support for values with commas in k1=v1,k2=[v2,v3,v4],k3=v3 into the Config Command

benstopford committed:
KAFKA-1464: Changed throttled-replica-list property from format "0-106:1-107"  to "0,106:1,107"

benstopford committed:
KAFKA-1464: Change ReassignPartitionCommand logic. Blocking option has been removed. It has the form:

kafka_reassign_partitions.sh … —execute —throttle 100 //executes the reassignment with the throttle, or updates the throttle value if the reassignment is already running

kafka_reassign_partitions.sh … —verify //checks to see if there is a throttle engaged and removes it if the execution has completed

benstopford committed:
KAFKA-1464: whitespace only

benstopford committed:
KAFKA-1464: small tidy up

benstopford committed:
KAFKA-1464: Flipped the order of the definition of throttled replicas to match PartitionId:BrokerId,PartitionId:BrokerId:...

apurvam commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -134,6 +135,8 @@ class ReplicaFetcherThread(name: String,
         trace("Follower %d set replica high watermark for partition [%s,%d] to %s"
           .format(replica.brokerId, topic, partitionId, followerHighWatermark))
+      if (quota.isThrottled(topicAndPartition))
+        quota.record(partitionData.toByteBufferMessageSet.sizeInBytes)
```
Nitpick: Would it be better to just use `messageSet` from line 118 instead of doing `partitionData.toByteBufferMessageSet` again? This way you save an allocation.

benstopford replied:
thanks

apurvam commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+  */
+class ReplicationQuotaManager(val config: ReplicationQuotaManagerConfig,
+                              private val metrics: Metrics,
+                              private val replicationType: String,
```
Would this be better as a typed enum rather than a string? Makes compile time checks stronger.

benstopford replied:
Yes. Agree.

apurvam commented on core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala:
```diff
@@ -0,0 +1,187 @@
+    info(s"Unthrottled took: $unthrottledTook, Throttled took: $throttledTook")
+    assertTrue(s"Unthrottled replication of ${unthrottledTook}ms should be < ${expectedDuration * 1000}ms",
+      unthrottledTook < expectedDuration * 1000)
+    assertTrue((s"Unthrottled replication of ${throttledTook}ms should be < ${expectedDuration * 1000}ms"),
```
This message doesn't match the test. Shouldn't it be "Throttled Replication of N ms should > M ms"

benstopford replied:
Thanks. Changed.

junrao commented on clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java:
```diff
@@ -111,7 +111,11 @@ public void record(double value, long timeMs) {
      * @param timeMs
      */
-    private void checkQuotas(long timeMs) {
+    public void checkQuotas() {
```
@param timeMs in the comment is no long valid.

benstopford replied:
thanks

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:
```diff
@@ -459,4 +460,62 @@ public double measure(MetricConfig config, long now) {
+
+        //When we record anything, if there is no time, rate will be infinite
+        record(rate, config, 1000);
+        assertEquals(1000, measure(rate, config),0);
```
The comment in line 470 doesn't seem to match the test. Also, space before 0.

benstopford replied:
good spot. thanks

junrao commented on clients/src/test/java/org/apache/kafka/common/metrics/MetricsTest.java:
```diff
@@ -459,4 +460,62 @@ public double measure(MetricConfig config, long now) {
+        record(rate, config, 1000);
+        assertEquals(1000, measure(rate, config),0);
+
+        //In the first window the rate is a fraction of the whole 1s window
```
Is this comment accurate?

benstopford replied:
yes it is. have clarified the test comment a bit.

junrao commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -38,7 +38,14 @@ import collection.Map
+  def fetchEntityConfig(zkUtils: ZkUtils,entityType: String, entityName: String): Properties
+}
+
+object AdminUtils extends Logging with AdminUtilities{
```
space before {

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -113,14 +122,19 @@ object ConfigCommand {
-    if (props.containsKey(LogConfig.MessageFormatVersionProp)) {
-      println(s"WARNING: The configuration ${LogConfig.MessageFormatVersionProp}=${props.getProperty(LogConfig.MessageFormatVersionProp)} is specified. " +
-        s"This configuration will be ignored if the version is newer than the inter.broker.protocol.version specified in the broker.")
+    if(opts.options.has(opts.addConfig)) {
```
space after if

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -145,27 +159,28 @@ object ConfigCommand {
             .withValuesSeparatedBy(',')
     val helpOpt = parser.accepts("help", "Print usage information.")
     val forceOpt = parser.accepts("force", "Suppress console prompts")
+    val jsonOpt = parser.accepts("json", "Configurations can be specified as JSON. This allows the comma separated lists to be passed as values.")
```
This doesn't seem to be used?

benstopford replied:
oops - thanks

junrao commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -191,8 +241,8 @@ object ReassignPartitionsCommand extends Logging {
       " Note that this only generates a candidate assignment, it does not execute it.")
-    val executeOpt = parser.accepts("execute", "Kick off the reassignment as specified by the --reassignment-json-file option.")
-    val verifyOpt = parser.accepts("verify", "Verify if the reassignment completed as specified by the --reassignment-json-file option.")
+    val executeOpt = parser.accepts("execute", "Kick off the reassignment as specified by the --reassignment-json-file option. If a reassignment is already running, the script will await completion then exit.")
```
Does --execute block?

benstopford replied:
No. Mistake. Thanks

junrao commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -211,7 +261,11 @@ object ReassignPartitionsCommand extends Logging {
                       .ofType(classOf[String])
     val disableRackAware = parser.accepts("disable-rack-aware", "Disable rack aware replica assignment")
-
+    val throttleOpt = parser.accepts("throttle", "The movement of partitions will be throttled to this value (B/s). Rerunning with this option, whilst a rebalance is in progress, will alter the throttle value.")
```
B/s => bytes/sec ?

benstopford replied:
done

junrao commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -221,7 +275,34 @@ object ReassignPartitionsCommand extends Logging {
   extends Logging {
-  def reassignPartitions(): Boolean = {
+
+  private def addThrottle(throttle: Long): Unit = {
```
addThrottle => maybeAddThrottle ?

benstopford replied:
done. thanks

junrao commented on core/src/main/scala/kafka/consumer/ConsumerFetcherThread.scala:
```diff
@@ -22,8 +22,7 @@ import kafka.cluster.BrokerEndPoint
 import kafka.common.{ErrorMapping, TopicAndPartition}
-import scala.collection.JavaConverters
-import JavaConverters._
+import scala.collection.{Map, JavaConverters}
```
unused import JavaConverters

benstopford replied:
done. thanks

junrao commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -111,7 +117,8 @@ class DelayedFetch(delayMs: Long,
     // Case D
-    if (accumulatedSize >= fetchMetadata.fetchMinBytes)
+    if (accumulatedSize >= fetchMetadata.fetchMinBytes
+      || (accumulatedThrottledSize >= fetchMetadata.fetchMinBytes && !quota.isQuotaExceeded()))
```
If quota is not exceeded, should we check (accumulatedSize + accumulatedThrottledSize) >= fetchMetadata.fetchMinBytes?

benstopford replied:
Yes. That's a great spot. Thank you.

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -18,14 +18,15 @@
 package kafka.server
 
 import java.util.Properties
+import java.util.concurrent.locks.{Lock, ReentrantReadWriteLock}
```
A few unused imports.

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -517,8 +524,12 @@ object KafkaConfig {
+  val NumReplicationQuotaSamplesDoc = "The number of samples to retain in memory for replication quotas"
+  val QuotaWindowSizeSecondsDoc = "The time span of each sample for client quotas"
+  val ReplicationQuotaWindowSizeSecondsDoc = "The time span of each sample for replication quotas"
+  val ThrottledReplicationRateLimitDoc = "A long representing the upper bound on replication traffic for replicas enumerated in the " +
```
the upper bound => the upper bound in bytes/sec ?

benstopford replied:
good spot

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -701,8 +714,10 @@ object KafkaConfig {
       .define(QuotaWindowSizeSecondsProp, INT, Defaults.QuotaWindowSizeSeconds, atLeast(1), LOW, QuotaWindowSizeSecondsDoc)
-
+      .define(ReplicationQuotaWindowSizeSecondsProp, INT, Defaults.ReplicationQuotaWindowSizeSeconds, atLeast(1), LOW, ReplicationQuotaWindowSizeSecondsDoc)
+      .define(ThrottledReplicationRateLimitProp, LONG, Defaults.ThrottledReplicationLimit, atLeast(0), MEDIUM, ThrottledReplicationRateLimitDoc)
```
Would it be better for ThrottledReplicationRateLimitProp to be Double, which matches the type of Quota?

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -41,7 +41,8 @@ import org.apache.kafka.common.requests.{ControlledShutdownResponse, ControlledS
 import org.apache.kafka.common.utils.AppInfoParser
 
-import scala.collection.mutable
+import scala.collection
```
unused import

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -267,17 +269,19 @@ class ReplicaFetcherThread(name: String,
-      if (partitionFetchState.isActive)
-        requestMap(new TopicPartition(topic, partition)) = new JFetchRequest.PartitionData(partitionFetchState.offset, fetchSize)
+    partitionMap.foreach { case ((partition, partitionFetchState)) =>
+      if (partitionFetchState.isActive && withinQuota(partition))
```
Would it be better to just check quota.isQuotaExceeded once and make the same decision for each throttled partition on whether it should be included or not?

benstopford replied:
sure thing.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -267,17 +269,19 @@ class ReplicaFetcherThread(name: String,
+        requestMap(new TopicPartition(partition.topic, partition.partition)) = new JFetchRequest.PartitionData(partitionFetchState.offset, fetchSize)
     }
-
+    trace("Created a fetch request for partitions " + partitionMap.keys.map(_.toString))
```
There is already a trace statement in AbstractFetcherThread that logs each fetch request. Do we still need this trace logging?

benstopford replied:
thanks. removed

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+import kafka.server.ReplicationQuotaManagerConfig._
+import kafka.utils.CoreUtils._
+import kafka.utils.Logging
+import org.apache.kafka.common.MetricName
```
Unused imports MetricName and Rate.

benstopford replied:
thanks

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+}
+
+object Constants {
+  val allReplicas = Seq[Int](-1)
```
AllReplicas instead?

benstopford replied:
ok. cool

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+    inWriteLock(lock) {
+      this.quota = quota
+      if (metrics.metrics.containsKey(rateMetricName))
+        metrics.metrics.get(rateMetricName).config(getQuotaMetricConfig(quota))
```
The above may not be 100% safe since the metric could be expired and removed between the two statements. It's safer if we save metrics.metrics.get(rateMetricName) to a local val and then check null and update the config.

benstopford replied:
I've changed this both here and also in the ClientQuotaManager (which had the same logic)

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+    */
+  override def isQuotaExceeded(): Boolean = {
+    try {
+      sensor.checkQuotas()
```
Since sensor() has side effect, it would be clearer if all references to sensor are sensor().

benstopford replied:
Yes - good call. Thanks

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+    } catch {
+      case qve: QuotaViolationException =>
+        if(logger.isTraceEnabled)
+          logger.trace("%s: Quota violated for sensor (%s), metric: (%s), metric-value: (%f), bound: (%f)".format(replicationType, sensor.name(), qve.metricName, qve.value, qve.bound))
```
The convention is to use trace() which does does the if check already. Also, for string formatting, we are moving towards the s notation instead of format. Ditto in a few other places.

benstopford replied:
Thanks for the heads up. Changed

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+  override def isThrottled(topicPartition: TopicAndPartition): Boolean = {
+    val partitions = throttledPartitions.get(topicPartition.topic)
+    if (partitions != null)
+      partitions.contains(topicPartition.partition) || partitions == allReplicas
```
partitions == allReplicas should probably be partitions eq allReplicas?

benstopford replied:
OK - makes sense.

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+    * @param partitions the set of throttled partitions
+    * @return
+    */
+  def markThrottled(topic: String, partitions: Seq[Int]): Unit = {
```
To be consistent, if there is no return value, we just do method() {}.

benstopford replied:
OK - thanks for the heads up - have changed tidied up where this was wrong elsewhere too.

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,192 @@
+    * @return
+    */
+  def markThrottled(topic: String, partitions: Seq[Int]): Unit = {
+    throttledPartitions.put(topic, partitions)
```
If partitions is empty, should we remove that topic from the map?

benstopford replied:
No, I missed that somehow. Have added it now.

junrao commented on core/src/test/scala/integration/kafka/api/ClientQuotasTest.scala:
```diff
@@ -132,9 +132,6 @@ class QuotasTest extends KafkaServerTestHarness {
 
     // Consumer should read in a bursty manner and get throttled immediately
     consume(consumers.head, numRecords)
-    // The replica consumer should not be throttled also. Create a fetch request which will exceed the quota immediately
```
Is there a reason to remove these? It seems the test is still useful.

benstopford replied:
How strange. That was removed when I merged from my previous branch. Good spot.

junrao commented on core/src/test/scala/unit/kafka/admin/AdminTest.scala:
```diff
@@ -31,9 +31,10 @@ import kafka.common.TopicAndPartition
+import kafka.utils.TestUtils._
 
 import scala.collection.{Map, immutable}
+import unit.kafka.admin.ReplicationQuotaUtils._
```
unused import

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+      0 -> Seq(100, 101)
+    ), servers = servers)
+
+    //Given throttle set so replication will take at least 10 sec (we won't wait this long)
```
10 sec seems inaccurate now?

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+    produceMessages(servers, topicName, numMessages, acks = 0, msgSize)
+    assertEquals(expectedDurationSecs, numMessages * msgSize / initialThrottle)
+
+    //Start rebalance (use a separate thread as it'll be slow)
```
Is this really done in a separate thread?

benstopford replied:
It certainly isn't! Thanks.

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+    val actual = zkUtils.getPartitionAssignmentForTopics(Seq(topicName))(topicName)
+    assertEquals(actual.values.flatten.toSeq.distinct.sorted, Seq(101, 102))
+
+    //Then command should have taken more than 1 second to complete as it is throttled
```
1 second seems inaccurate now?

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+
+    //Then command should have taken more than 1 second to complete as it is throttled
+    assertTrue(s"Expected replication to be > ${expectedDurationSecs * 0.9 * 1000} but was $took", took > expectedDurationSecs * 0.9 * 1000)
+    assertTrue(s"Expected replication to be < 20s but was $took", took < expectedDurationSecs * 2 * 1000)
```
20s seems inaccurate now?

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/admin/ReplicationQuotaUtils.scala:
```diff
@@ -0,0 +1,50 @@
+  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+  * specific language governing permissions and limitations under the License.
+  */
+package unit.kafka.admin
```
remove unit

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaManagerQuotasTest.scala:
```diff
@@ -0,0 +1,154 @@
+    expect(log.logEndOffset).andReturn(20L).anyTimes()
+    expect(log.logEndOffsetMetadata).andReturn(new LogOffsetMetadata(20L)).anyTimes()
+
+    //if we ask for len 100 return a message
```
Is len 100 or 1?

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/server/ReplicationQuotaManagerTest.scala:
```diff
@@ -0,0 +1,124 @@
+import kafka.common.TopicAndPartition
+import kafka.server.QuotaType._
+import kafka.server.{ReplicationQuotaManager, ReplicationQuotaManagerConfig}
+import org.apache.kafka.common.metrics.stats.{Rate, SimpleRate}
```
unused imports

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/server/ReplicationQuotaManagerTest.scala:
```diff
@@ -0,0 +1,124 @@
+    assertFalse(quota.isThrottled(TopicAndPartition("MyOtherTopic", 0)))
+  }
+
+  def tp1(id: Int): TopicAndPartition = new TopicAndPartition("topic1", id)
```
Could this and next method be private?

benstopford replied:
sure can

junrao commented on core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala:
```diff
@@ -0,0 +1,187 @@
+  def shouldMatchQuotaReplicatingThroughAnAsymmetricTopology(leaderThrottle: Boolean): Unit = {
+    /**
+      * In short we have 8 brokers, 2 are not-started. We put replicas for all partitions on the
+      * not-started brokers, so when we start them we can monitor replication from the 6 to the 2.
```
Do we "put replicas for all partitions on the not-started brokers"?

benstopford replied:
clarified. thanks

junrao commented on core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala:
```diff
@@ -0,0 +1,187 @@
+    info(s"Unthrottled took: $unthrottledTook, Throttled took: $throttledTook")
+    assertTrue(s"Unthrottled replication of ${unthrottledTook}ms should be < ${expectedDuration * 1000}ms",
+      unthrottledTook < expectedDuration * 1000)
+    assertTrue((s"Unthrottled replication of ${throttledTook}ms should be < ${expectedDuration * 1000}ms"),
```
The text in here and line 172 seem inaccurate.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -536,10 +541,18 @@ class ReplicaManager(val config: KafkaConfig,
+              var fetch = log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //If the partition is marked as throttled, and we are over-quota then exclude it
+              if (quota.isThrottled(TopicAndPartition(topic, partition)) && quota.isQuotaExceeded) //TODO - could also check ISR inclusion??
```
With this, after recording the bytes, the quota could be exceeded? The earlier approach where we use  quota.isQuotaExceed(expectedBytes) seems more conservative and is less likely for quota to be exceeded. Is there a reason not to use quota.isQuotaExceed(expectedBytes)?

benstopford replied:
Yes - the idea seemed like a good one, but it led to some complexities. The main problem was that the ratio of the RequestSize:QuotaSize correlated with the amount the throttle would be undercut. Also, if you asked for more than the quota in a single request you could never make progress. The upshot was that the behaviour was a little unintuitive, and the simpler mechanism seems to work well on aggregate. 
So after thinking about it for a while I decided to ditch the isExceededBy(bytes) approach. Keep it simple ... and this way it more closely matches the way the follower works more closely. I hope that makes sense.

junrao commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -95,8 +97,12 @@ class DelayedFetch(delayMs: Long,
               } else if (fetchOffset.messageOffset < endOffset.messageOffset) {
-                // we need take the partition fetch size as upper bound when accumulating the bytes
-                accumulatedSize += math.min(endOffset.positionDiff(fetchOffset), fetchStatus.fetchInfo.fetchSize)
+                // we take the partition fetch size as upper bound when accumulating the bytes (skip if a throttled partition)
```
It seems that we need to handle case C better. If the follower is lagging on old segments and the quota is exceeded, we may return an empty result well before max wait. Returning an empty result early occasionally is fine. However, in this case, it seems that this can happen continuously.

benstopford replied:
Good spot. I think this should be as simple as:

```
   if (!(quota.isThrottled(topicAndPartition) && quota.isQuotaExceeded()))
         return forceComplete()
```

but i'll need to write a test which will take a little time.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -489,8 +493,9 @@ class ReplicaManager(val config: KafkaConfig,
+
       val fetchMetadata = FetchMetadata(fetchMinBytes, fetchOnlyFromLeader, fetchOnlyCommitted, isFromFollower, fetchPartitionStatus)
-      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, responseCallback)
+      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota ,responseCallback)
```
There could be a subtle issue with timeout. Say we check the quota and it's exceeded, and we will put the DelayedFetch in the purgatory. If no more bytes are produced, the DelayedFetch has to wait for maxWait. However, quota could become available before maxWait (and we won't get a chance to check). 

One potential way to address this is that if quota is exceeded, we calculate the amount of time that needs to pass before quota is available. We set the timeout in DelayedFetch to the be smaller of that time and maxWait. Then, if DelayedFetch expires, in DelayedFetch.onComplete(), if minBytes is not satisfied and maxWait hasn't been exceeded. We put DelayedFetch to purgatory again.

benstopford replied:
I did consider this issue. You are right that we could make the leader algorithm more responsive by altering the timeout (I didn't think of that - good idea). 
For now I'm inclined to raise a jira for this as a future enhancement. It shouldn't affect throttling significantly. The extra delay will even out over time.  
Also, this problem exists on the follower too so the optimisation is only of value for the leader side of the throttle. 
The main concern I have is actually the lack of smarts on the follower, particularly if throttled partitions enter the ISR, as the follower logic is very basic.

junrao replied:
Yes, the follower has a similar issue. There is already logic to add a delay per partition. So, if a partition is throttled in the follower, we can calculate a delay from quota and delay the partition accordingly. We probably also need to change the backoff logic in AbstractFetcherThread a bit. Instead of always backing off for a fixed amount of time, it's probably better to backoff based on the smallest delay among all partitions.

I agree that this is probably not a common issue. It only becomes a big issue if the maxWait or replica backoff time are configured very large (say close to the metric window \* sample size). So, we can address that in a followup jira.

benstopford committed:
KAFKA-1464: moved QuotaType back to being an Enumeration (of sorts)

benstopford committed:
KAFKA-1464: Range of small tidy up tasks. One functional change to improve the resiliancy of the metrics config update in the Client/ReplicationQuotaManagers, which may not have been 100% safe, since the metric could be expired and removed between the two statements. Changed to use a local variable and null check to avoid this issue.

benstopford committed:
KAFKA-1464: Stylistic changes only, based on PR comments

ijuma commented on core/src/main/scala/kafka/utils/Logging.scala:
```diff
@@ -32,8 +32,9 @@ trait Logging {
-    if (logger.isTraceEnabled())
+    if (logger.isTraceEnabled()) {
       logger.trace(msgWithLogIdent(msg))
+    }
```
Probably unintended change?

benstopford replied:
thanks

ijuma commented on core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala:
```diff
@@ -93,7 +126,7 @@ object ReassignPartitionsCommand extends Logging {
   }
 
-  def generateAssignment(zkUtils: ZkUtils, brokerListToReassign: Seq[Int], topicsToMoveJsonString: String, disableRackAware: Boolean): (Map[TopicAndPartition, Seq[Int]], Map[TopicAndPartition, Seq[Int]]) = {
+  def generateAssignment(zkUtils: ZkUtils, brokerListToReassign: Seq[Int], topicsToMoveJsonString: String, disableRackAware: Boolean): (Map[TopicAndPartition, Seq[Int]], Map[TopicAndPartition, Seq[Int]]) =  {
```
Probably unintended change?

benstopford replied:
thanks

benstopford committed:
KAFKA-1464: Fixes issue where the throttled replica list was not being deleted when the corresponding topic config was removed. Also includes reference to another bug where the the throttled replica list is not propagated to the quota manager if added on topic creation (not a core use case but a bug none the less)

benstopford committed:
KAFKA-1464: Further to previous commit, ensure the throttled replica list is removed if "" is passed in the property

benstopford committed:
KAFKA-1464: Remove redundant TODO only

benstopford committed:
KAFKA-1464: Replaced two lines of test code lost during merge.

benstopford committed:
KAFKA-1464: remove unused imports, fixed comments, made methods private (no functional change)

benstopford committed:
KAFKA-1464: whitespace/brackets only

benstopford committed:
KAFKA-1464: Fix checkstyle formatting error only

junrao commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -72,21 +102,49 @@ object ClientConfigOverride {
+class BrokerConfigHandler(private val brokerConfig: KafkaConfig, private val quotaManagers: QuotaManagers) extends ConfigHandler with Logging {
+  def processConfigChanges(brokerId: String, properties: Properties) {
+    if (brokerConfig.brokerId == brokerId.trim.toInt) {
+      val limit = if (properties.containsKey(ThrottledReplicationRateLimitProp)) properties.getProperty(ThrottledReplicationRateLimitProp).toLong else Defaults.ThrottledReplicationLimit
```
processConfigChanges() only gets called if there is overridden config on brokerId in ZK. So, if that doesn't exist, it seems that we won't apply the static default ThrottledReplicationLimit in broker property file?

junrao replied:
I think we can just get rid of the static config ThrottledReplicationRateLimitProp and just rely on the dynamic broker level config, which is more flexible.

benstopford replied:
Yes - I raised a bug for this, but it's not a big deal so lets lose the config it as you say.

benstopford replied:
The nice thing about having it in the config is validation. If we remove the prop, we'd probably need another list of dynamic broker configs somewhere. What do you think?

junrao replied:
Yes, we are doing similar things in KIP-55. We are deprecating the static client quota configs in the broker in favor of dynamic quotas. So, for new broker configs, if it can be made dynamically, it seems it's less confusing to also add a static config in the broker property.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -132,6 +132,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
 
   var kafkaHealthcheck: KafkaHealthcheck = null
   var metadataCache: MetadataCache = null
+  var quotaManagers:  QuotaFactory.QuotaManagers = null
```
extra space after :

junrao commented on core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala:
```diff
@@ -0,0 +1,198 @@
+    val msg = msg100KB
+    val msgCount: Int = 1000
+    val expectedDuration = 10 //Keep the test to N seconds
+    var throttle: Int = msgCount * msg.length / expectedDuration
```
throttle should be type long?

benstopford replied:
thanks. done

junrao commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,203 @@
+    } catch {
+      case qve: QuotaViolationException =>
+        if(logger.isTraceEnabled)
+          logger.trace("%s: Quota violated for sensor (%s), metric: (%s), metric-value: (%f), bound: (%f)".format(replicationType, sensor().name(), qve.metricName, qve.value, qve.bound))
```
convention: use trace().

benstopford replied:
thanks. done

junrao commented on core/src/main/scala/kafka/utils/CoreUtils.scala:
```diff
@@ -20,6 +20,7 @@ package kafka.utils
 import java.io._
 import java.nio._
 import java.nio.channels._
+import java.util.Properties
```
unused import

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/admin/AdminTest.scala:
```diff
@@ -385,45 +385,111 @@ class AdminTest extends ZooKeeperTestHarness with Logging with RackAwareTest {
+      AdminUtils.changeTopicConfig(server.zkUtils, topic, new Properties)
+      checkConfig(Defaults.MaxMessageSize, Defaults.RetentionMs, Defaults.ThrottledReplicasList,  quotaManagerIsThrottled = false)
+
+      //Add config bakc
```
typo bakc

benstopford replied:
thanks

junrao commented on core/src/test/scala/unit/kafka/admin/AdminTest.scala:
```diff
@@ -385,45 +385,111 @@ class AdminTest extends ZooKeeperTestHarness with Logging with RackAwareTest {
+      checkConfig(maxMessageSize, retentionMs, "0:0,1:0,2:0", quotaManagerIsThrottled = true)
+
+      //Now ensure updating to "" removes the throttled replica list
+      AdminUtils.changeTopicConfig(server.zkUtils, topic, new Properties(){put(LogConfig.ThrottledReplicasListProp, "")})
```
Perhaps we should just test excluding the property, which is consistent with the test in line 484?

benstopford replied:
We have that already on line 428. Are you concerned about it being there, or objecting to the format of the test?

junrao commented on core/src/test/scala/unit/kafka/admin/AdminTest.scala:
```diff
@@ -385,45 +385,111 @@ class AdminTest extends ZooKeeperTestHarness with Logging with RackAwareTest {
+    }
+
+    try {
+      val limit = 42
```
Should we make limit long since ThrottledReplicationRateLimitProp is of type long?

benstopford replied:
ok

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+    val actual = zkUtils.getPartitionAssignmentForTopics(Seq(topicName))(topicName)
+    assertEquals(actual.values.flatten.toSeq.distinct.sorted, Seq(101, 102))
+
+    //Then command should have take longer than the throttle rate
```
take => taken

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+    val initialThrottle: Long = 1000 * 1000
+    produceMessages(servers, topicName, numMessages = 200, acks = 0, valueBytes = 100 * 1000)
+
+    //Start rebalance (use a separate thread as it'll be slow)
```
Are we using a separate thread?

junrao commented on core/src/test/scala/unit/kafka/admin/ReassignPartitionsClusterTest.scala:
```diff
@@ -96,14 +95,96 @@ class ReassignPartitionsClusterTest extends ZooKeeperTestHarness with Logging {
+    //Check throttle config
+    checkThrottleConfigAddedToZK(initialThrottle, servers, topicName)
+
+    //Now re-run the same assignment with a larger throttle, which should only act to increase the throttle and make progress (again use a thread so we can check ZK whilst it runs)
```
Are we using a separate thread?

junrao commented on core/src/test/scala/unit/kafka/log/LogConfigTest.scala:
```diff
@@ -65,6 +64,34 @@ class LogConfigTest {
+    assertFalse(isValid("100:0,10:   "))
+  }
+
+  def isValid(configValue: String): Boolean = {
```
Should this be private?

junrao commented:
I left the following comment on SimpleRate earlier.
Having two different rates is going to make it harder for developers to decide which one to use. If this is strictly better than Rate, perhaps we should just change Rate.windowSize(). If this is just for testing, perhaps we can create SimpleRate in test?

benstopford committed:
KAFKA-1464: Altered reassign command to only throttle replicas that are either move sources or move destinations.

benstopford committed:
KAFKA-1464: Formatting, typos, etc based on Jun's feedback

benstopford committed:
KAFKA-1464: Formatting, typos, etc based on Jun's feedback part 2

benstopford committed:
KAFKA-1464: Formatting, typos, etc based on Jun's feedback part 3

benstopford committed:
KAFKA-1464: Reverting something that wasn't supposed to go in.

benstopford committed:
KAFKA-1464: whitespace

benstopford committed:
KAFKA-1464: Added throttling support for Old Segments (sheepishly tested manually,  will add automated tests later)

benstopford committed:
KAFKA-1464: Added test to ensure that we definitely throttle requests from old segments. The chnage for this was in DelayedFetch, in the previous commit.

benstopford committed:
KAFKA-1464: Renamed throttle config (naming error). Tidied doc.

benstopford committed:
KAFKA-1464: Renamed again

apurvam commented on core/src/main/scala/kafka/server/ReplicationQuotaManager.scala:
```diff
@@ -0,0 +1,202 @@
+    *
+    * @param quota
+    */
+  def updateQuota(quota: Quota) {
```
You should log a message here, with information of the old and new quota topic for the broker/topic/partition being modified. Otherwise these dynamic changes without any auditing will be impossible to debug.

benstopford replied:
It should be visible from the logging in the DynamicConfigManager (line 106). Do you not see that?

apurvam replied:
Yes, I see it now. Sorry for the false alarm.

benstopford committed:
KAFKA-1464: Output text for the reassign command only.

benstopford committed:
KAFKA-1464: Removed whitespace / triggered build

junrao commented:
Thanks for the patch. LGTM. We can address the remaining minor issues in a followup jira.
